cmd/snap-confine/mount-support.c

```diff
@@ -62,47 +62,82 @@
 // TODO: fold this into bootstrap
 static void setup_private_mount(const char *snap_name)
 {
-	char tmpdir[MAX_BUF] = { 0 };
+	// TODO: remove after switching to global umask handling.
+	mode_t old_mask = umask(0);
 
-	// Create a 0700 base directory, this is the base dir that is
-	// protected from other users.
+	// Create a 0700 base directory, this is the base dir that is protected
+	// from other users. This directory name is NOT randomly generated. This
+	// has several properties:
+	//
+	// Users can relate to the name and can find the temporary directory as
+	// visible from within the snap. If this directory was random it would be
+	// harder to find because there may be situations in which multiple
+	// directories related to the same snap name would exist.
+	//
+	// Snapd can partially manage the directory. Specifically on snap remove
+	// snapd could remove the directory and everything in it, potentially
+	// avoiding runaway disk use on a machine that either never reboots or uses
+	// persistent /tmp directory.
 	//
-	// Under that basedir, we put a 1777 /tmp dir that is then bind
-	// mounted for the applications to use
-	sc_must_snprintf(tmpdir, sizeof(tmpdir), "/tmp/snap.%s_XXXXXX", snap_name);
-	if (mkdtemp(tmpdir) == NULL) {
-		die("cannot create temporary directory essential for private /tmp");
+	char base_dir[MAX_BUF] = { 0 };
+	char tmp_dir[MAX_BUF] = { 0 };
+	int base_dir_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	int tmp_dir_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	sc_must_snprintf(tmp_dir, sizeof(tmp_dir), "/tmp/snap.%s/tmp",
+			 snap_name);
+	sc_must_snprintf(base_dir, sizeof(base_dir), "/tmp/snap.%s", snap_name);
+
+	// Create /tmp/snap.$SNAP_NAME/ 0700 root.root
+	if (mkdir(base_dir, 0700) < 0 && errno != EEXIST) {
+		die("cannot create base directory %s", base_dir);
 	}
-	// now we create a 1777 /tmp inside our private dir
-	mode_t old_mask = umask(0);
-	char *d = sc_strdup(tmpdir);
-	sc_must_snprintf(tmpdir, sizeof(tmpdir), "%s/tmp", d);
-	free(d);
-
-	if (mkdir(tmpdir, 01777) != 0) {
-		die("cannot create temporary directory for private /tmp");
+	base_dir_fd = open(base_dir,
+			   O_RDONLY | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+	if (base_dir_fd < 0) {
+		die("cannot open base directory %s", base_dir);
 	}
-	umask(old_mask);
-
+	if (fchmod(base_dir_fd, 0700) < 0) {
+		die("cannot chmod base directory %s to 0700", base_dir);
+	}
+	if (fchown(base_dir_fd, 0, 0) < 0) {
```

Comment:
Wouldn't it be more careful to abort execution if the `base_dir` is not owned by `root:root` in the first place? Since the directory now has a fixed name an attacker can prepare dirs there that already contain prepared content, or even mounted file systems (like FUSE).

Also there is a race condition between the `open()` and the `fchmod()` / `fchown()`. If the directory already exists and is owned by a regular user then this regular user might be able to rename the directory before ownership is actually transferred to root.

Reply:
To mount a filesystem the attacker would have to be root and we are not trying to defend against that.

As for the race condition, indeed it might be better to abort if the directory, as opened, is owned by non-root user. I will look at a way to avoid the race condition if I can and if not, just give up early as you suggested.

Reply:
Re the race, this isn't a problem because the parent directory is owned by root, not the user, so normal DAC prevents a non-root users from renaming while we've still got the fd. We aren't protecting from unconfined root cause if you have that you can just replace snapd.

Reply:
> Re the race, this isn't a problem because the parent directory is owned by root, not the user, so normal DAC prevents a non-root users from renaming while we've still got the fd. We aren't protecting from unconfined root cause if you have that you can just replace snapd.

I beg to differ. `base_dir` is `/tmp/snap.%s` and `base_dir_fd` refers to that directory. So we can have the following situation:

<pre>
attacker $ mkdir /tmp/snap.mysnap
# snap-confine opens /tmp/snap.mysnap as `base_dir_fd`
attacker $ mv /tmp/snap.mysnap /tmp/my.special.dir
# snap-confine calls fchown() on `base_dir_fd`
</pre>

So as a result `/tmp/my.special.dir` would now be owned by root. This is giving an attacker the possibility to bring arbitrarily named directories into existence that are owned by root (could even be moved to a home directory if not located on a different file system).

Also the actual `/tmp/snap.mysnap` will not be the expected root-owned directory, leading to unspecified results in follow up logic in snap-confine.

Reply:
I will be changing this in a follow-up fix today.

Reply:
> I beg to differ. `base_dir` is `/tmp/snap.%s` and `base_dir_fd` refers to that directory.

Yes, but I was thinking this happened after we shared the mount namespace, where in the parent dir, /tmp, is not 1777. I may have been wrong in thinking that, but it doesn't matter, @zyga is looking to address this.

Reply:
This took a while to be something I can propose: https://github.com/snapcore/snapd/pull/6714